prov/verbs/src/verbs_eq.c

```diff
@@ -575,24 +582,33 @@ fi_ibv_eq_cm_process_event(struct fi_ibv_eq *eq,
 				return ret;
 		}
 		ep = container_of(fid, struct fi_ibv_ep, util_ep.ep_fid);
-		if (fi_ibv_is_xrc(ep->info))
-			return fi_ibv_eq_xrc_connected_event(eq, cma_event,
-							     entry, len, acked);
+		if (fi_ibv_is_xrc(ep->info)) {
+			fastlock_acquire(&eq->lock);
+			ret = fi_ibv_eq_xrc_connected_event(eq, cma_event,
+							    entry, len, acked);
+			fastlock_release(&eq->lock);
+			return ret;
+		}
 		entry->info = NULL;
 		break;
 	case RDMA_CM_EVENT_DISCONNECTED:
 		ep = container_of(fid, struct fi_ibv_ep, util_ep.ep_fid);
 		if (fi_ibv_is_xrc(ep->info)) {
+			fastlock_acquire(&eq->lock);
 			fi_ibv_eq_xrc_disconnect_event(eq, cma_event, acked);
+			fastlock_release(&eq->lock);
```

Comment:
Does the locking need to stay outside the event handler function?

Reply:
No, it can easily move within the called function if that is preferred. I was just trying to keep it simple. This function is only called from here and exists to avoid complaints about code complexity.

> Does the locking need to stay outside the event handler function?
No, it can easily move within the called function if that is preferred. I was just trying to keep it simple. This function is only called from here and exists to avoid complaints about code complexity.

Reply:
My only thought was that moving it inside the function would keep `fi_ibv_eq_cm_process_event` smaller. But feel free to leave it here if that makes it more readable.

Reply:
Ok, thanks!